### PR TITLE
Add event-based volume changes

### PR DESCRIPTION
**Pull request by ElitCoder**
Starting with
$ ./minidsp-volume-steps-rs event <event path>
listens for events on the specified path.

Currently, the volume key "u" and "d" are mapped to volume up and
volume down.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -50,25 +55,28 @@ fn apply_gain(gain: &str) {
         .expect("Failed to update gain");
 }
 
+fn gain_changed(current: f64, new: f64) -> bool {
+    // Stringify to miniDSP format too see if the gain has changed
```
Comment (Lundez) [inline src/main.rs:63]: Do we really need to stringify? How about `abs(current - new) >= 0.1`?

File: src/main.rs [hunk 2]
```diff
@@ -50,25 +55,28 @@ fn apply_gain(gain: &str) {
...
     let final_gain = clamp(scaled_gain, MINIDSP_GAIN_MIN, MINIDSP_GAIN_MAX);
 
-    // Stringify to miniDSP format too see if the gain has changed
-    let old_gain_string = format!("{:.1}", current);
-    let new_gain_string = format!("{:.1}", final_gain);
-    let gain_changed = old_gain_string != new_gain_string;
-
-    if gain_changed {
-        apply_gain(&new_gain_string);
+    let changed = gain_changed(current, final_gain);
```
Comment (Lundez) [inline src/main.rs]: `gain_changed` was a better name imo. By returning to that we can also remove the comment at output. 🥇

File: src/main.rs [hunk 3]
```diff
@@ -93,3 +101,64 @@ fn main() {
...
+    // Retrieve current gain
+    let current_gain = get_gain();
+    let mut new_gain = current_gain;
+    if increase {
+        new_gain += GAIN_OFFSET;
+    } else {
+        new_gain -= GAIN_OFFSET;
+    }
+    // Clamp to make sure the gain is in the correct range
+    let clamped_gain = clamp(new_gain, MINIDSP_GAIN_MIN, MINIDSP_GAIN_MAX);
```
Comment (Lundez) [inline src/main.rs]: We call `camp(...)` twice, every time using `MINIDSP_GAIN_MIN, MINIDSP_GAIN_MAX`. I think that should be removed and simply used directly.

File: src/main.rs [hunk 4]
```diff
@@ -93,3 +101,64 @@ fn main() {
...
+    if gain_changed(current_gain, clamped_gain) {
+        apply_gain(&format!("{:.1}", clamped_gain));
+    }
+}
+
+fn act_on_event(event: evdev_rs::InputEvent) {
+    const VOLUME_UP_KEY: evdev_rs::enums::EV_KEY = evdev_rs::enums::EV_KEY::KEY_U;
```
Comment (Lundez) [inline src/main.rs:125]: as simple as _what the fuck_ am I reading <3 

Looks very complicated but I understand it hahaha

File: src/main.rs [hunk 5]
```diff
@@ -93,3 +101,64 @@ fn main() {
...
+    if event.is_type(&evdev_rs::enums::EventType::EV_KEY) && event.value == 1 {
+        match event.event_code {
+            evdev_rs::enums::EventCode::EV_KEY(VOLUME_UP_KEY) => change_gain(true),
+            evdev_rs::enums::EventCode::EV_KEY(VOLUME_DOWN_KEY) => change_gain(false),
+            _ => (), // Ignore everything else
+        }
+    }
+}
+
+fn event(path: String) {
```
Comment (Lundez) [inline src/main.rs:138]: Why are we having a file of events? Is it to save the latest gain?

**Review by Lundez — APPROVED** [review]
See comments, just clearing up code. Had no energy to deep-dive on why things are like they are or why change is done 😄

File: src/main.rs (hunk 1 above)
Comment (ElitCoder) [reply src/main.rs:63]: Technically no, but the minidsp binary makes me scared... an invalid gain here might result in max volume which my ears won't like :smile:

File: src/main.rs (hunk 4 above)
Comment (ElitCoder) [reply src/main.rs:125]: I think it's beautiful :heart:

File: src/main.rs (hunk 5 above)
Comment (ElitCoder) [reply src/main.rs:138]: It's both a stream and a file, think of e.g. /dev/urandom. The "event-file" is opened like a normal file to receive events.

**Comment by Lundez** [comment]
Let's say LGTM 👌👌👌👌

**Pull request merged**